### PR TITLE
[6_2_X][TIMOB-1028] Android: Implement AudioRecorder (Parity)

### PR DESCRIPTION
**Pull request by ypbnv**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-1028

**Description:**
Cherry pick of: https://github.com/appcelerator/titanium_mobile/pull/9067

### CODE REVIEW

File: android/modules/media/src/java/ti/modules/titanium/media/AudioRecorderProxy.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	@Kroll.method
+	public void start() {
+		tiAudioRecorder.startRecording();
+	}
+
+	@Kroll.method
+	public TiFileProxy stop() {
+		return new TiFileProxy(TiFileFactory.createTitaniumFile(tiAudioRecorder.stopRecording(),false));
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/AudioRecorderProxy.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/MediaModule.java [hunk 2]
```diff
@@ -1285,6 +1308,13 @@ public boolean getIsCameraSupported()
 
 	}
 
+	@Kroll.method
+	@Kroll.getProperty
+	public boolean getCanRecord()
+	{
+		return TiApplication.getInstance().getPackageManager().hasSystemFeature( "android.hardware.microphone");
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/MediaModule.java]: remove whitespace after `(`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	//Byte array used for reading from the native buffer to the output stream
+	private byte[] audioData;
+	private int bufferSize = 0;
+	private File tempFileReference;
+	private FileOutputStream fileOutputStream = null;
+	private AudioRecord audioRecord;
+
+	public TiAudioRecorder() {
+		//Get the minimum buffer size according to the device recording capabilities
+		bufferSize = AudioRecord.getMinBufferSize(RECORDER_SAMPLE_RATE,RECORDER_CHANNELS,RECORDER_AUDIO_ENCODING);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		return recording;
+	}
+
+	public boolean isStopped() {
+		return stopped;
+	}
+
+	public void startRecording() {
+		try {
+			tempFileReference =  TiFileHelper.getInstance().getTempFile(AUDIO_RECORDER_TEMP_FILE,true);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+			fileOutputStream = new FileOutputStream(tempFileReference);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+
+		//Initialize the audio recorded with big enough buffer to ensure smooth reading from it without overlap
+		audioRecord = new AudioRecord(MediaRecorder.AudioSource.MIC,RECORDER_SAMPLE_RATE, RECORDER_CHANNELS,RECORDER_AUDIO_ENCODING, bufferSize*4);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+		if (audioRecord != null) {
+			int recordState = audioRecord.getState();
+			if (recordState == 1) {
+				audioRecord.stop();
+			}
+			audioRecord.setRecordPositionUpdateListener(null);
+			audioRecord.release();
+			audioRecord = null;
+			try {
+				resultFile = TiFileHelper.getInstance().getTempFile(AUDIO_RECORDER_FILE_EXT_WAV,true);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+				createWaveFile(resultFile.getAbsolutePath());
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		}
+		return resultFile != null ? resultFile.getAbsolutePath():"";
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space before and after `:`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+			audioRecord.stop();
+		}
+	}
+
+	public void resumeRecording() {
+		//Guard for calling resume before starting the recording
+		if (audioRecord != null) {
+			paused = false;
+			audioRecord.startRecording();
+			audioRecord.read(audioData,0,bufferSize);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+		try {
+			//set the result file output stream
+			out = new FileOutputStream(outFilename);
+			in = new FileInputStream(tempFileReference);
+			totalAudioLen = in.getChannel().size();
+			totalDataLen = totalAudioLen + 36;
+
+			//Write the WAV header
+			writeWaveFileHeader(out, totalAudioLen, totalDataLen,RECORDER_SAMPLE_RATE, channels, byteRate);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private void writeWaveFileHeader(FileOutputStream out, long totalAudioLen, long totalDataLen,
+									 long longSampleRate, int channels, long byteRate) throws IOException {
+		byte[] header = new byte[44];
+
+		header[0] = 'R'; // RIFF/WAVE header
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: could use an array?

```Java
byte[] header = {
    'R', 'I', 'F'
    ...
};
```

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+		header[36] = 'd';
+		header[37] = 'a';
+		header[38] = 't';
+		header[39] = 'a';
+		header[40] = (byte) (totalAudioLen & 0xff);
+		header[41] = (byte) ((totalAudioLen >> 8) & 0xff);
+		header[42] = (byte) ((totalAudioLen >> 16) & 0xff);
+		header[43] = (byte) ((totalAudioLen >> 24) & 0xff);
+
+		out.write(header, 0, 44);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]:
```Java
44 -> header.length
```

File: android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+		@Override
+		public void onMarkerReached(AudioRecord recorder) {
+
+		}
+
+		@Override
+		public void onPeriodicNotification(AudioRecord recorder) {
+			try {
+				audioRecord.read(audioData,0,bufferSize);
```
Comment (garymathews) [inline android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java]: space after `,`

**Review by garymathews — CHANGES_REQUESTED** [review]
See above; these changes should also be pushed to `master`.

**Comment by lokeshchdhry** [comment]
FR Passed.

With proper permissions, the audio recorder's record, pause, play, stop works as expected.

1. Checked by recording audio, stopping & playing it.
2. Recording audio, pausing , recording again , stopping & playing.

Studio Ver:     4.9.1.201707200100
SDK Ver:        6.2.0.v20170801014621
OS Ver:         10.12.3
Xcode Ver:      Xcode 8.3.3
Appc NPM:       4.2.9
Appc CLI:       6.2.3-21
Ti CLI Ver:     5.0.14
Alloy Ver:      1.9.13
Node Ver:       6.10.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Nexus 5 --- Android 6.0.1

**Comment by lokeshchdhry** [comment]
Will merge when the requested changes are done.

**Comment by ypbnv** [comment]
@garymathews Updated.

**Review by garymathews — APPROVED** [review]
CR: **PASS**

**Pull request merged**